lib/ci-helper.ts

```diff
@@ -182,18 +191,44 @@ export class CIHelper {
                 meta.commitInGitGit = undefined;
                 result = true;
             }
+            bases.add(`-p${info.baseCommit}`);
+            heads.add(`-p${info.headCommit}`);
+        }
+
+        if (heads.size > 0) {
+            /*
+             * Generate throw-away octopus merges to combine multiple commit
+             * ranges into a single one.
+             */
+            const octopus = async (set: Set<string>): Promise<string> => {
+                const array = Array.from(set);
+                if (array.length === 1) {
+                    return array[0];
+                }
+                return await
+                    git(["commit-tree", ...array, emptyTreeName, "-m", "()"],
+                        { workDir: this.workDir });
+            };
 
+            const range1 = `${await octopus(bases)}..${await octopus(heads)}`;
+            const range2 =
+                "refs/remotes/upstream/maint~100..refs/remotes/upstream/pu";
+            const start = Date.now();
             const out = await git(["-c", "core.abbrev=40", "range-diff", "-s",
-                                   info.baseCommit, info.headCommit,
-                                   "refs/remotes/upstream/pu"],
+                                   range1, range2],
                                   { workDir: this.workDir });
+            const duration = Date.now() - start;
+            if (duration > 2000)
+                console.log(`warning: \`git range-diff ${
+                    range1} ${range2}\` took ${
+                    duration / 1000} seconds`);
             for (const line of out.split("\n")) {
                 const match =
                     line.match(/^[^:]*: *([^ ]*) [!=][^:]*: *([^ ]*)/);
                 if (!match) {
                     continue;
                 }
-                const messageID2 = match[1] === info.headCommit ? messageID :
+                const messageID2: string | undefined =
```

Comment:
The type declaration could be removed (not present on line 178).

Reply:
You're right!